### DISCUSSION TITLE
Working day management in formula

### DISCUSSION THREAD

**Discussion by MPE04** [freeplane/freeplane#1708] [Help] [answered]
Hello,
I try to use freeplane for project management by creating various formula. At the time, I want to calculate the estimate end date for a task. I've got a formula that's work for calendar days : `= new Date().plus(node['RAF'].num0.toInteger())`

Is there any possibility to perform the same but for working day ? I mean, without taking saturday and sunday into account.

**Comment by euu2021** [comment, marked as answer]
Is RAF an attribute where the user sets the number of days to complete the task? If so, then:

```
import java.time.LocalDate
import java.time.DayOfWeek

// Function to add working days to the current date
def addWorkingDays(startDate, daysToAdd) {
    LocalDate result = startDate
    int addedDays = 0
    while (addedDays < daysToAdd) {
        result = result.plusDays(1)
        // Check if the current day is a weekday
        if (!(result.dayOfWeek == DayOfWeek.SATURDAY || result.dayOfWeek == DayOfWeek.SUNDAY)) {
            addedDays++
        }
    }
    return result
}

// Main script starts here
// Retrieve the 'RAF' attribute value as an integer
int rafDays = node['RAF'].num0.toInteger()

// Calculate the start date, assuming today. Adjust as necessary.
LocalDate startDate = LocalDate.now()

// Calculate the estimated end date excluding weekends
LocalDate estimatedEndDate = addWorkingDays(startDate, rafDays)

// Output the estimated end date
estimatedEndDate.toString()
```

Adjust the Start Date: The script currently uses the current date as the start date (LocalDate.now()). If you need to start from a different date, you'll need to modify the startDate initialization accordingly.

Then, there are the holidays. Do you care about the holidays?

(created using [FP script GPT](https://github.com/euu2021/training-Freeplane-script-GPT))

**Reply by MPE04** [reply]
OK so it should be a script, not a formula.

I update as follow to put the result in a specific attribute : 

`import java.time.LocalDate
import java.time.DayOfWeek

// Function to add working days to the current date
def addWorkingDays(startDate, daysToAdd) {
    LocalDate result = startDate
    int addedDays = 0
    while (addedDays < daysToAdd) {
        result = result.plusDays(1)
        // Check if the current day is a weekday
        if (!(result.dayOfWeek == DayOfWeek.SATURDAY || result.dayOfWeek == DayOfWeek.SUNDAY)) {
            addedDays++
        }
    }
    return result
}

// Main script starts here
// Retrieve the 'RAF' attribute value as an integer
int rafDays = node['RAF'].num0.toInteger()

// Calculate the start date, assuming today. Adjust as necessary.
LocalDate startDate = LocalDate.now()

// Calculate the estimated end date excluding weekends
LocalDate estimatedEndDate = addWorkingDays(startDate, rafDays)

// Output the estimated end date
if (node.attributes.containsKey("endEstimate") == true){
	node.attributes.set("endEstimate",estimatedEndDate.toString())
}
else {
	node.attributes.add("endEstimate",estimatedEndDate.toString())
}`

I don't care about the holiday, so the code works for me ! Thanks !

**Reply by euu2021** [reply]
> OK so it should be a script, not a formula.

It can be a formula. Put a `=` at the beginning, and it becomes a formula.
